### PR TITLE
update AIDE init tasks for configurable timeout

### PR DESCRIPTION
**Pull request by shepdelacreme**

### CODE REVIEW

**Comment by shepdelacreme** [comment]
@jamescassell Please let me know if this accomplishes what you need. You can set `rhel7stig_aide_init_timeout` to a sufficiently high value and then `rhel7stig_wait_for_aide_init: false` so the playbook won't check status.

File: tasks/prelim.yml [hunk 1]
```diff
@@ -243,11 +243,32 @@
...
+        async: "{{ rhel7stig_aide_init_timeout | default(omit) }}"
+        poll: "{{ (rhel7stig_aide_init_timeout is defined) | ternary(0,omit) }}"
+        register: rhel7stig_aide_init_status
+        when: not rhel7stig_aide_db_status.stat.exists
+
+      - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Notify AIDE init handler"
+        command: "true"
+        when:
+          - rhel7stig_aide_init_status.changed
+          - rhel7stig_wait_for_aide_init
```
Comment (jamescassell) [inline tasks/prelim.yml]: I'd just move this line to the handler itself.

File: tasks/prelim.yml [hunk 2]
```diff
@@ -243,11 +243,32 @@
...
+        command: /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_db_file }}'
+        async: "{{ rhel7stig_aide_init_timeout | default(omit) }}"
+        poll: "{{ (rhel7stig_aide_init_timeout is defined) | ternary(0,omit) }}"
+        register: rhel7stig_aide_init_status
+        when: not rhel7stig_aide_db_status.stat.exists
+
+      - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Notify AIDE init handler"
+        command: "true"
+        when:
+          - rhel7stig_aide_init_status.changed
```
Comment (jamescassell) [inline tasks/prelim.yml]: Move the notify to the previous task instead.

**Comment by jamescassell** [comment]
After sleeping on it, I think it is probably best to wait to run the init until after all the other STIG remediation tasks are complete.  This is to prevent aide from taking a snapshot of files that will be changed by later steps in the the role.

**Comment by shepdelacreme** [comment]
Noted. I moved it back to be a handler but added in the options to have it run in multiple modes:

- Fire and forget by setting the polling interval to 0
- blocking/sync mode by setting timeout to 0
- "async" mode by setting timeout and polling interval to > 0 (this is the default mode although it operates the same as sync mode

Also noticed that we probably shouldn't be writing directly to /var/lib/aide/aide.db.gz so added logic to write to /var/lib/aide/aide.db.new.gz then move it to /var/lib/aide/aide.db.gz unless you are running the init task in "fire and forget" mode then it just writes it directly.

**Comment by shepdelacreme** [comment]
Once this is merged I think we should add notify to the tasks that edit /etc/aide.conf as well so that a new baseline db is generated with the updated settings.

File: tasks/prelim.yml [hunk 3]
```diff
@@ -249,26 +249,20 @@
...
-        register: rhel7stig_aide_init_status
-        when: not rhel7stig_aide_db_status.stat.exists
-
-      - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Notify AIDE init handler"
-        command: "true"
-        when:
-          - rhel7stig_aide_init_status.changed
-          - rhel7stig_wait_for_aide_init
-        notify: init aide status
+      - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Set AIDE database output file"
```
Comment (jamescassell) [inline tasks/prelim.yml]: There's no need for this task.  Always write to the temp file.  Worst case, put it in defaults or vars.

File: handlers/main.yml [hunk 4]
```diff
@@ -23,13 +23,20 @@
...
+  async: "{{ rhel7stig_aide_init_timeout | default(omit) }}"
+  poll: "{{ rhel7stig_aide_init_interval | default(omit) }}"
+  when: not rhel7stig_aide_db_status.stat.exists
+  notify: move aide db
+  tags:
+    - aide
+
+- name: move aide db
+  command: "mv {{ rhel7stig_aide_db_output_file }} {{ rhel7stig_aide_db_file }}"
+  when:
```
Comment (jamescassell) [inline handlers/main.yml]: Do this unconditionally.  AIDE init will continue writing to the file originally opened, even if the file changes fs location underneath it.  This is how the RHEL6-STIG role works.  (That role still has the 'killed after 45 seconds' problem, though.)

File: handlers/main.yml [hunk 5]
```diff
@@ -23,13 +23,20 @@
...
 
-- name: init aide status
-  async_status:
-    jid: "{{ rhel7stig_aide_init_status.ansible_job_id }}"
-  register: rhel7stig_aide_init_job_result
-  until: rhel7stig_aide_init_job_result.finished
-  retries: 60
-  delay: 5
+- name: init aide
+  command: /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_db_output_file }}'
```
Comment (jamescassell) [inline handlers/main.yml]: Probably want to move the aide init to after any other handlers that make filesystem changes.

File: tasks/prelim.yml [hunk 6]
```diff
@@ -249,26 +249,20 @@
...
             name: aide
             state: present
+        notify: init aide
 
       - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Check for existing AIDE database"
         stat:
           path: "{{ rhel7stig_aide_db_file }}"
         register: rhel7stig_aide_db_status
         check_mode: no
+        changed_when: not rhel7stig_aide_db_status.stat.exists
```
Comment (jamescassell) [inline tasks/prelim.yml]: Perfect!

**Comment by jamescassell** [comment]
Is it desirable to have 'aide --check' return clean after running this role and a reboot?

File: handlers/main.yml (hunk 5 above)
Comment (shepdelacreme) [reply handlers/main.yml]: Hmm is this as simple as moving these handlers to the end of the file?

Comment (jamescassell) [reply handlers/main.yml]: handlers are always executed in the order specified in the handlers file, regardless of the order they were notified.

File: handlers/main.yml [hunk 7]
```diff
@@ -56,3 +39,17 @@
...
+  poll: "{{ rhel7stig_aide_init_interval | default(omit) }}"
+  when: not rhel7stig_aide_db_status.stat.exists
+  notify: move aide db
+  tags:
+    - aide
+
+- name: move aide db
+  command: "mv {{ rhel7stig_aide_temp_db_file }} {{ rhel7stig_aide_db_file }}"
+  tags:
+    - aide
```
Comment (jamescassell) [inline handlers/main.yml]: Please have valid lines.

File: handlers/main.yml [hunk 8]
```diff
@@ -56,3 +39,17 @@
...
   command: auditctl -f {{ rhel7stig_auditd_failure_flag }}
   failed_when: no
+
+- name: init aide
+  command: /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_temp_db_file }}'
+  async: "{{ rhel7stig_aide_init_timeout | default(omit) }}"
+  poll: "{{ rhel7stig_aide_init_interval | default(omit) }}"
+  when: not rhel7stig_aide_db_status.stat.exists
+  notify: move aide db
+  tags:
```
Comment (jamescassell) [inline handlers/main.yml]: Is there a need for tags in the handlers file?

File: handlers/main.yml [hunk 9]
```diff
@@ -56,3 +39,17 @@
...
+    - aide
+
+- name: move aide db
+  command: "mv {{ rhel7stig_aide_temp_db_file }} {{ rhel7stig_aide_db_file }}"
+  tags:
```
Comment (jamescassell) [inline handlers/main.yml]: handlers behave as if they had
```
tags:
- always
```

File: handlers/main.yml (hunk 5 above)
Comment (shepdelacreme) [reply handlers/main.yml]: We should probably move the reboot handler to the bottom of the file then.

File: handlers/main.yml (hunk 8 above)
Comment (shepdelacreme) [reply handlers/main.yml]: There used to be a need but it looks like they fixed that.

File: handlers/main.yml (hunk 5 above)
Comment (jamescassell) [reply handlers/main.yml]: True, though I don't think we use the reboot handler.

**Comment by shepdelacreme** [comment]
Just tested running `aide --check` after a reboot and it comes back unclean but only because of the its noticing the ansible async job file is missing.
```
---------------------------------------------------
Removed files:
---------------------------------------------------

removed: /root/.ansible_async/228982313278.15010.tmp
```

**Comment by jamescassell** [comment]
If we get rid of async and always do it synchronously, or with the true fire-and-forget mode I posted we won't get that.

**Comment by shepdelacreme** [comment]
I'm fine doing it synchronously but I need a way to ensure the task finishes before shutting the box down.

**Comment by jamescassell** [comment]
Presumably it will usually finish before the default SSH timeout of 10 minutes.  Otherwise, the play will fail and you will know.  In my experience, even if it doesn't finish within the SSH timeout period, the aide init still finishes successfully, eventually.

**Comment by shepdelacreme** [comment]
Yes most of the time. However if I'm running this with packer or other similar image builders it is going to shut the machine down immediately after the playbook finishes. There is the potential for the task to just not finish unless the shutdown command would wait for the aide init process to finish before shutting down.

**Comment by shepdelacreme** [comment]
How about this then? sync/blocking or fire and forget only. Fire and forget uses your shell nohup command...accomplished with two different handlers set using a var.

File: tasks/prelim.yml [hunk 10]
```diff
@@ -243,11 +243,21 @@
...
+            name: aide
+            state: present
+        notify: init aide
+
+      - name: "PRELIM | RHEL-07-020030 RHEL-07-020040 RHEL-07-021600 RHEL-07-021610 RHEL-07-021620 | Check for existing AIDE database"
+        stat:
+          path: "{{ rhel7stig_aide_db_file }}"
+        register: rhel7stig_aide_db_status
+        check_mode: no
+        changed_when: not rhel7stig_aide_db_status.stat.exists
```
Comment (jamescassell) [inline tasks/prelim.yml]: Since we're conditionally notifying the handler here, we don't need the "when" on the handler itself.

File: handlers/main.yml [hunk 11]
```diff
@@ -46,3 +39,16 @@
 - name: update running audit failure mode
   command: auditctl -f {{ rhel7stig_auditd_failure_flag }}
   failed_when: no
+
+- name: init aide and wait
+  command: /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_temp_db_file }}'
+  when: not rhel7stig_aide_db_status.stat.exists
```
Comment (jamescassell) [inline handlers/main.yml]: we can get rid of this line now.

File: handlers/main.yml [hunk 12]
```diff
@@ -46,3 +39,16 @@
...
+  notify: move aide db
+
+- name: init aide
+  shell: nohup /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_temp_db_file }}' > /dev/null &
+  when: not rhel7stig_aide_db_status.stat.exists
```
Comment (jamescassell) [inline handlers/main.yml]: and this line

**Comment by jamescassell** [comment]
I like your new approach.

File: tasks/prelim.yml (hunk 10 above)
Comment (shepdelacreme) [reply tasks/prelim.yml]: Good catch

**Comment by shepdelacreme** [comment]
The thought of backing up an old/existing DB has crossed my mind. I guess it somewhat depends how people use this role. I generally just run it on a new/clean box at the beginning of the build process. Are people running this role on existing machines? On a regular schedule, like every month or something?

I suppose we could have a var:
`rhel7stig_force_init_aide_db: true`
That would for AIDE init regardless of the outcome of the `stat` task.

Then add a task after the `stat` task that moves the existing DB file to `/var/lib/aide/aide.db.{{ datatime }}.gz` or something. However I don't want to bother adding all that logic if no one really cares and is fine just overwriting the existing DB.

**Comment by jamescassell** [comment]
I'm fine over-writing the DB.  We only do it when something changed that would necessitate re-creating it anyway.

**Comment by shepdelacreme** [comment]
So just a force or overwrite db variable? I think I can just drop that on the stat task. Should be easy.

**Comment by jamescassell** [comment]
maybe a default "rhel7stig_allow_init_aide_db: true" then change the handler "when" to be "when: rhel7stig_allow_init_aide_db"

**Comment by jamescassell** [comment]
Or better yet, just put it on the mv task and call it rhel7stig_allow_overwrite_aide_db

File: tasks/fix-cat2.yml [hunk 13]
```diff
@@ -897,6 +897,7 @@
         regexp: '([A-Z]+ = .*)(sha256\+sha512)(.*)'
         replace: '\1sha512\3'
       register: rhel_07_021620_sha256_512_result
+      notify: "{{ rhel7stig_aide_handler }}"
       failed_when:
         - rhel_07_021620_sha256_512_result | failed
         - rhel_07_021620_sha256_512_result.rc != 257
```
Comment (jamescassell) [inline tasks/fix-cat2.yml]: While we're making changes here, we can get rid of the custom failed_when since this is now tied to the PRELIM task.  Could be a separate PR, though.

File: tasks/fix-cat2.yml [hunk 14]
```diff
@@ -907,11 +908,13 @@
         regexp: '([A-Z]+ = .*)(sha256)(.*)'
         replace: '\1sha512\3'
       register: rhel_07_021620_sha256_result
+      notify: "{{ rhel7stig_aide_handler }}"
       failed_when:
```
Comment (jamescassell) [inline tasks/fix-cat2.yml]: same here

File: tasks/prelim.yml [hunk 15]
```diff
@@ -249,7 +249,7 @@
         yum:
             name: aide
             state: present
-        notify: init aide
+        notify: "{{ rhel7stig_aide_handler }}"
```
Comment (jamescassell) [inline tasks/prelim.yml]: Good.

File: handlers/main.yml [hunk 16]
```diff
@@ -42,13 +42,12 @@
...
   notify: move aide db
 
 - name: init aide
   shell: nohup /usr/sbin/aide --init -B 'database_out=file:{{ rhel7stig_aide_temp_db_file }}' > /dev/null &
-  when: not rhel7stig_aide_db_status.stat.exists
   notify: move aide db
 
 - name: move aide db
   command: "mv {{ rhel7stig_aide_temp_db_file }} {{ rhel7stig_aide_db_file }}"
+  when: not rhel7stig_aide_db_status.stat.exists or rhel7stig_overwrite_aide_db
```
Comment (jamescassell) [inline handlers/main.yml]: Looks good.

File: tasks/fix-cat2.yml (hunk 13 above)
Comment (shepdelacreme) [reply tasks/fix-cat2.yml]: There isn't a stat task for /etc/aide.conf in prelim yet so the failed_when is technically still needed for now. I'll add that in a later PR I think.

Comment (jamescassell) [reply tasks/fix-cat2.yml]: later PR is fine.  /etc/aide.conf is provided by the aide package, so it's fine to error out if it's not there.

**Comment by jamescassell** [comment]
Awesome! Ship it!
